Source/Core/Core/HW/GPFifo.h

```diff
@@ -16,29 +21,44 @@ constexpr u32 GATHER_PIPE_PHYSICAL_ADDRESS = 0x0C008000;
 constexpr u32 GATHER_PIPE_SIZE = 32;
 constexpr u32 GATHER_PIPE_EXTRA_SIZE = GATHER_PIPE_SIZE * 16;
 
-// Init
-void Init();
-void DoState(PointerWrap& p);
-
-// ResetGatherPipe
-void ResetGatherPipe();
-void UpdateGatherPipe();
-void CheckGatherPipe();
-void FastCheckGatherPipe();
-
-bool IsBNE();
-
-// Write
-void Write8(u8 value);
-void Write16(u16 value);
-void Write32(u32 value);
-void Write64(u64 value);
-
-// These expect pre-byteswapped values
-// Also there's an upper limit of about 512 per batch
-// Most likely these should be inlined into JIT instead
-void FastWrite8(u8 value);
-void FastWrite16(u16 value);
-void FastWrite32(u32 value);
-void FastWrite64(u64 value);
+class alignas(GATHER_PIPE_SIZE) GPFifoManager final
+{
+public:
+  // Init
+  void Init();
+  void DoState(PointerWrap& p);
+
+  // ResetGatherPipe
+  void ResetGatherPipe();
+  void UpdateGatherPipe(Core::System& system);
```

Comment:
Maybe pass a referecnce of the `System` instance into the GPFifoManager on construction, so that the interface itself doesn't need to repeatedly pass the system into it.